### ISSUE TITLE
Document project interlinking

### ISSUE THREAD

**Issue by glasnt** [pybee/pybee.github.io#121]
Things depend on each other, and we don't have a good picture of this interconnectivity documented at the moment.

**Comment by goanpeca** [comment]
@glasnt sort of creating a dependency graph for all the tools, to produce a nice image?

**Comment by freakboy3742** [comment]
Agreed that this is definitely something where we are lacking. We need to revisit the homepage and project descriptions to give a better overall feel of how the project pieces fit together.

**Comment by VeldaKiara** [comment]
Would you be open to giving more context to help

**Comment by VeldaKiara** [comment]
Looking into this, for the sprint session today

**Comment by VeldaKiara** [comment]
Hey @freakboy3742 

Here is the first draft I have, and I would love to get suggestions for it.

I had a few tweaks from @kattni and Baptiste

![Image](https://github.com/user-attachments/assets/7dfc775c-701c-45ae-983f-efa1c86d84ea)

**Comment by freakboy3742** [comment]
So - apologies if I've led you astray here, but I'm not sure this is what is needed.

This is a nice graphical layout of the same project hierarchy that the website uses, but it doesn't really add any *new* information that isn't already conveyed by the website.

What is currently missing is an explanation of how all the pieces fit together. What's the relationship between Toga and Colosseum? What's the connection between Podium and Rubicon ObjC? 

What we need is something that visualises relationships *between* the parts, not just enumerating the parts. I was thinking of [this sort of diagram of the Android stack](https://developer.android.com/static/guide/platform/images/android-stack_2x.png)  - something that shows how different pieces are layered on top of each other. 

So - in terms of how the apps work at runtime, the apps are all at the top, sitting on top of Toga; Toga sits on Colosseum, which builds on Travertino (but Toga also sits on Travertino directly; Colosseum is an "alpha" project that isn't used actively); on macOS and iOS, Toga also sits on Rubicon ObjC. On GTK, Toga depends on PyGObject (which we don't maintain); on Windows, Toga depends on Python.net (which we don't maintain); on Android, Toga depends on Chaquopy (which isn't a BeeWare project, but shares a maintainer in Malcolm); on Web, Toga depends on PyScript (which we don't maintain); and on consol, Toga depends on Textual (which we don't maintain).

Then there's the packaging story. Each app is an output of Briefcase; and Briefcase depends on Python Apple Support (on iOS and macOS), Chaquopy (on Android), Linux system packges, and the official CPython Windows embedded build.

Now... how you put all that into a single diagram - especially the app vs deployment split.... I don't know :-)

Lastly - to be consistent with brand guidelines, any diagram headings should be using BeeWare's identity font - Cutive - and should be leaning toward sans-serif for longer text blocks.

**Comment by VeldaKiara** [comment]
Yes this actually helps, I also might have neglected to mention that this was the first part the second part would be adding information on the gutter on the side to share about the relationship. But I can figure out a way to show the relationship more using the above context

**Comment by johnzhou721** [comment]
![Image](https://github.com/user-attachments/assets/56505073-9e27-463e-8986-1208c31ed2af)

@freakboy3742 Is this maybe a good proof of concept of what you're having in mind? This is very rough, and will need to be digitized in some way (maybe with LaTeX and friends, or I might find PowerPoint easier), and I've probably gotten a bunch of stuff wrong.

**Comment by freakboy3742** [comment]
@johnzhou721 That's certainly closer to what I had in mind - although I'd definitely be looking for a much clearer "block with single name" layout for each piece, and less "annotations" on the side. Using different colors/shading would also be a given, as would digitising the image in a format that is widely editable - SVG would be my pick (definitely not Powerpoint, as it's a proprietary tool that many, including myself, don't have).

However, I'd also suggest that maybe you should leave this ticket to the person who opened it. You've already got multiple projects underway; I'd suggest focussing your attention on those, rather than adding another project that someone else is already working on.

**Comment by johnzhou721** [comment]
<img width="1331" alt="Image" src="https://github.com/user-attachments/assets/0d0b3f33-fb08-4f50-b0b1-ea0fa9a51064" />

Nice suggestion. I've already took a stab on this -- using SVG and converted to PDF. Turns out we came up with some of the same things (``less annotations``) while I'm working. I don't use PowerPoint unless for school projects either (just realized that, it was a leftover memory from when I used to see my mom make flowcharts using it).

Shading is not done *yet*. Regarding multiple projects -- sorry I saw this too late and already did some initial work.

Question: Any trademarks BeeWare / Chaquopy has registered and any branding guidelines you have that would prevent me from putting this on my own repo (I will denote that it is UNOFFICIAL, though).

**Comment by johnzhou721** [comment]
Oh sorry I misread comment I thought less annotations in the blocks but you said on the side... Will need to simplify.

**Comment by freakboy3742** [comment]
I also said "maybe don't work on a ticket that someone else has already said they're working on"...

**Comment by johnzhou721** [comment]
Right... I am indeed embarrased from spending so much time and duplicating effort

You said ``leave this ticket to the person who opened it`` in your last comment but glasnt isn't active... do you mean the person who sprinted on this?

**Comment by freakboy3742** [comment]
> You said `leave this ticket to the person who opened it` in your last comment but glasnt isn't active... do you mean the person who sprinted on this?

Yes - I mean the person who was working on this issue just 4 days ago.

**Comment by johnzhou721** [comment]
FWLIW (for whatever little it's worth): I won't work on this anymore, but since I hate showing stuff without sources and I hate sources with too much clutter, I spend 1 minute making some minor tweaks and posted my diagram at https://github.com/johnzhou721/beeware-struct, but I'm abandoning it now since @VeldaKiara was the person who started work on this.

**Comment by VeldaKiara** [comment]
Based  on this information and inspiration from the attached image, I came up with this. 

I added the Toga Mascot to show that the dependencies are for Toga, and on the runtime dependencies, the ones with different colours are maintained by Beeware.

Let me know if this closer to what you had in mind and modifications to be made.
I have attached with mascots and without.



![Image](https://github.com/user-attachments/assets/7192cf1b-eb67-432b-8e2e-563ec639c642)

![Image](https://github.com/user-attachments/assets/5647ff08-4bfd-4de8-924d-099cff463ee8)

> So - apologies if I've led you astray here, but I'm not sure this is what is needed.
> 
> This is a nice graphical layout of the same project hierarchy that the website uses, but it doesn't really add any _new_ information that isn't already conveyed by the website.
> 
> What is currently missing is an explanation of how all the pieces fit together. What's the relationship between Toga and Colosseum? What's the connection between Podium and Rubicon ObjC?
> 
> What we need is something that visualises relationships _between_ the parts, not just enumerating the parts. I was thinking of [this sort of diagram of the Android stack](https://developer.android.com/static/guide/platform/images/android-stack_2x.png) - something that shows how different pieces are layered on top of each other.
> 
> So - in terms of how the apps work at runtime, the apps are all at the top, sitting on top of Toga; Toga sits on Colosseum, which builds on Travertino (but Toga also sits on Travertino directly; Colosseum is an "alpha" project that isn't used actively); on macOS and iOS, Toga also sits on Rubicon ObjC. On GTK, Toga depends on PyGObject (which we don't maintain); on Windows, Toga depends on Python.net (which we don't maintain); on Android, Toga depends on Chaquopy (which isn't a BeeWare project, but shares a maintainer in Malcolm); on Web, Toga depends on PyScript (which we don't maintain); and on consol, Toga depends on Textual (which we don't maintain).
> 
> Then there's the packaging story. Each app is an output of Briefcase; and Briefcase depends on Python Apple Support (on iOS and macOS), Chaquopy (on Android), Linux system packges, and the official CPython Windows embedded build.
> 
> Now... how you put all that into a single diagram - especially the app vs deployment split.... I don't know :-)
> 
> Lastly - to be consistent with brand guidelines, any diagram headings should be using BeeWare's identity font - Cutive - and should be leaning toward sans-serif for longer text blocks.

**Comment by VeldaKiara** [comment]
I get it now, define the relationships better through the diagram

**Comment by johnzhou721** [comment]
> This is a lot closer in terms of "brand guideline compliance" - but as with the previous version, I'm not sure it conveys any new information that isn't already conveyed by the hierarchy on the website.

I’d like to slightly disagree — it does convey that the higher level projects depend on the lower level ones even though it’s got a couple of bugs (travertino and coloss might not depend on rubicon-objc etc and briefcase’s position) — that said that is partly my mistake for I used a diagram like this except more compactly packed — I had not had any structure besides color to identify platforms so I apologize for misleading y’all.

**Comment by johnzhou721** [comment]
> Here is the updated image
> 
> ![Image](https://github.com/user-attachments/assets/d3be99ed-6858-4e35-af07-bfa8582fe0c2)

Very good documentation!

However, I think going as far as the kernel would be too low.

Stack the Toga backends *on top* of the PyScript etc. to show that they depend.

Might want to make horizontal bars grouping the Toga backends together. I see that is already on the bottom diagram; it'd be better if we integrated the platform-specific things into a big diagram instead of all those sections.

It'd be nice if we could show that all these stuff depends on the Python interpreter packaged by Python-Apple-Support etc. but I guess it's not going to be a strict requirement to show deps on the Python Interpreter.

Remember: Make higher level things *depend* on lower levels.

Just my two cents. Let's wait for @freakboy3742 for more feedback on this before you do another rework.

John

**Comment by freakboy3742** [comment]
So... this is better... but it's still not right. 

I have to apologize from the outset, because what I'm trying to explain here is simultaneously "obvious to me" (because I've spent a career looking at these sort of architecture explanation diagrams), and "not at all obvious" because what BeeWare is doing is... complex. The mapping of BeeWare's architecture to a diagram *isn't* simple... if it was, I would have added this diagram years ago.

That said: this diagram isn't there yet. John has highlighted some of the issues:

* It goes into *way* more detail than it needs to in some areas. We don't have any involvement with the XNU kernel. BeeWare's interaction with macOS ends at "we use macOS APIs". We don't need to document anything deeper than that unless it clarifies something about the relationship between the rest of BeeWare's tools - and in this case, it won't.
* This style of architecture diagram has a very specific "geometry". It's quite literally "high level" and "low level" - high level things are at the top, low level things are at the bottom. Having "native widgets" above "Python app", or "testing and automation" between "apps" and "briefcase" doesn't make sense in that world.
* It separates *some* of the vertical slices... but not all of them. It separates macOS from Linux, but not iOS from macOS - despite the fact that they're completely separate platforms. It doesn't separate PyScript or Textual... and sure, those run on macOS and Linux and Windows... but from *BeeWare's* perspective, they're different platforms - so they're almost certainly different vertical "channels" that talk to a low level "Browser" or "terminal API".
* Some concepts are on the diagram multiple times. Each OS is mentioned in a "stack"... but then is also mentioned in the bottom "Toga app" stack". Chaquopy is mentioned twice. 

Those sorts of discrepancies seem to come from a misunderstanding about the type of diagram I'm describing. It might be worth doing some research around this - an image search for "software block architecture diagram" will turn up dozens of this sort of image. Some have arrows, some don't; some are more detailed than others - but fundamentally, they're all following the same sort of pattern - how do the parts work together to build a whole. And, most importantly, they're always standalone - there aren't five of them for a single project.

But - the more importantly - we need to look at *why* are we adding this diagram? 

The problem we're trying to solve here isn't "the docs don't have a diagram". It's "the docs don't adequately explain how all the pieces fit together". BeeWare as project as a whole has a lot of subprojects. If you're a first timer looking at BeeWare, it's unlikely you'll understand why all these projects exist. *That* is what we're trying to clarify. Each time we add a box to the diagram, we need to ask:

* What is this box saying? 
* Why does it need to exist? 
* What relationship does it establish that wasn't previously clear? 
* Is *here* the right place for that box?
* Would this box be better represented as line between existing boxes?

Maybe a better place to start here is to describe a diagram *just* for Toga. That will cover a lot of the complexity, but it's a smaller scale problem so we can narrow down how a diagram of this nature should be laid out; once we have that diagram, we can add work out how to add the other complexities.

**Comment by VeldaKiara** [comment]
Thank you for the above feedback, this is progress.I will take the notes and try to come up with something we can iterate over.
